### ISSUE TITLE
Remove FocusRepo

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#2395]
### Vision statement / What / Requirements
`FocusRepo` is complex and hard to change: is there a simpler solution? In theory, the Android framework should be doing focus handling for us but the `FocusRepo` seems to reimplement a lot of this framework functionality (disclaimer: I'm not that familiar with the code). I'd like to explore removing it and replacing it with something simpler.

### Impact
Without making this change:
- we may continue to suffer from obscure focus bugs
- changing the focus handling code will be slow, challenging, and may result in regressions (disclaimer: I'm not that familiar with the code)

By making this change:
- we'll undo the issues above
- we may introduce regressions
- for all we know, the replacement could be harder to maintain (but I think not)

### Acceptance criteria
- [ ] FocusRepo is removed
- [ ] Focus behavior is equivalent to, or better than, existing behavior
  - [ ] This solution is simpler than FocusRepo
  - [ ] Go through open focus bugs: are they fixed with these changes? Are they more complicated?
- [ ] If applicable, any telemetry built around focus is working correctly

---

I wrote a proof-of-concept ([here is the branch](https://github.com/mcomella/firefox-tv/tree/host-poc-rm-focusrepo)) that removes FocusRepo without adding additional code. Testing on an emulator, there are two discrepancies from the existing codebase:
- `NavigationOverlay` is not focused when opened
- Moving up from nav url bar to page actions incorrectly focuses the "Exit Firefox" button

These seem mostly straightforward to fix to me so, assuming there are not many more significant focus issues, I don't think the complexity of the 350+ lines I removed is necessary.

---

I noticed that we have a few `<requestFocus` tags in XML that steal focus in some unexpected situations: removing them may simplify our focus handling code. See my explanation here for more: https://github.com/mozilla-mobile/firefox-tv/pull/2455#pullrequestreview-262617788

---

I do not consider this an aspirational refactor (which we don't file) because the proof-of-concept immediately shows the benefits of these changes and I think it's more than aspirational: I think we should do it next time we need to change focus code.

**Comment by mcomella** [comment]
@Baron-Severin @psymoon Do you have any thoughts on this, given my findings from the proof-of-concept?

**Comment by psymoon** [comment]
Hey @mcomella, I support this 💯 . Here is the list of major problems `FocusRepo` was addressing:
1. Pocket error state focus handling (error button vs. recyclerView) - https://github.com/mozilla-mobile/firefox-tv/issues/1130
2. ToolBar button focus handling (leftmost active order) - https://github.com/mozilla-mobile/firefox-tv/issues/906
3. Channel RecyclerView focus logic - https://github.com/mozilla-mobile/firefox-tv/issues/1178, https://github.com/mozilla-mobile/firefox-tv/pull/2177/commits/e0617ecea6aeaa1571376c4203b12b76882a7320 & https://github.com/mozilla-mobile/firefox-tv/pull/2177/commits/5daeac55f8997e1fe4d5d3d77570d1fc4d5bafbd (required for default focus logic for channels)
4. on screen change, focus gets lost (overlay - https://github.com/mozilla-mobile/firefox-tv/issues/2008, webrender - https://github.com/mozilla-mobile/firefox-tv/issues/1940#issuecomment-482699025)
5. Overlay losing focus when all tiles are deleted (https://github.com/mozilla-mobile/firefox-tv/issues/1146)

Notable minor issues: https://github.com/mozilla-mobile/firefox-tv/issues/2247, https://github.com/mozilla-mobile/firefox-tv/issues/2332

There may be a few other minor ones that I am missing; feel free to add more edge cases :)

**Comment by Baron-Severin** [comment]
>     1. Pocket error state focus handling (error button vs. recyclerView) - #1130

Number 1 should be a non-issue now that we've moved to channels, but +1 on all of the rest.  I'll add more if any come to mind, but at first glance that seems to cover it.

**Comment by athomasmoz** [comment]
Marking as a P1 to have the initial discussion. Making the recommended change that comes out of that discussion should be a separate ticket that can be prioritized

**Comment by AndiAJ** [comment]
Hi, I've re-tested this matter on **Firefox for FireTV v4.2 LAT2** using a **Fire TV Stick 2nd Gen** based on these remarks https://github.com/mozilla-mobile/firefox-tv/issues/2395#issuecomment-499960027

1. ToolBar button focus handling (leftmost active order) - #906
I've noticed that there's inconsistency in some situations while navigating from the Pinned tiles items to the Pocket items (it looks like the transition isn't made to the correct item in some cases).
Please review the [Navigation Pinned tiles items to the Pocket items](https://drive.google.com/open?id=1mhGfKq97DGXK301phmIyGnSmpKt3RKGe) VS [News & Politics to Pinned Tiles ](https://drive.google.com/open?id=1ejvWnXP0HwCqyzbzjXgFG3r5PXONONPW)

2. Channel RecyclerView focus logic - #1178, e0617ec & 5daeac5 (required for default focus logic for channels)
After removing all the Pinned tiles the container actually disappears automatically and can't focus on it.

3. On screen change, focus gets lost (overlay - #2008, webrender - #1940 (comment))

• "intermittent grey screen" aka lost EngineView focus
• "100% grey screen issue" aka WebRenderFragment fail to render

In both cases I've managed to reproduce the issues, but please note that **these happen quite fast** (a second or so)
Don't know previously how persistent they were.

4. Overlay losing focus when all tiles are deleted (#1146)
Blocked by https://github.com/mozilla-mobile/firefox-tv/issues/2398


Minor issues
Clicking left d-pad from urlbar moves to Pinned Tiles #2247
No issues - Focus is not changed 

Focus handling for turbo mode button #2332
No issues - Focus is not changed
